### DISCUSSION TITLE
OTA manually?

### DISCUSSION THREAD

**Discussion by gnome4711** [ESPresense/ESPresense#250] [General]
Hi guys,
u
is it possible to update (in my case downgrade) the firmware ota?
Since the 2.017 seems to be buggy I would like go back to the previous version. My ESP32's are not easy to reach so I don't want to connect every device to the PC.

Thanks in advance

gnome

**Comment by k3vyn32** [comment]
I would also be interested in how to do this.

**Comment by mk-maddin** [comment]
Would be interested, too - with old esp32-mqtt-room project I could simply flash a new Version via platformio, but even with Arduino OTA enabled I cannot Upload the bin File :/
Even tried to create a custom DNS entry for github to fake the latest version update but without any success

**Comment by mk-maddin** [comment]
ok - for everyone interested - got downgrade via OTA working :D

0. Install platform via command: pip3 install platformio 
1. ensure you have "Arduino OTA Update (default: ☐)" checkbox on the webpage enabled
![2022-02-20_16-08-18](https://user-images.githubusercontent.com/46523240/154849289-13d9d863-9ef5-498a-a9f5-7d4176804597.jpg)
2. RESTART the module e.g. via http://<ES32-IP>/restart or via power plug - without the restart the OTA functionality is not active!
3. Now we are downloading the source_code.zip (from whichever version required to downgrade) & the esp32.bin.
    We are creating a hidden subfolder of "./ESPresense-<VERSION>/.pio/build/esp32" and saving the esp32.bin in there as firmware.bin
    Now we can upload the firmware.bin via platformio:
    
    I have complete all of step 3 via following bash commands:

```
version="2.0.24"
bin_url="https://github.com/ESPresense/ESPresense/releases/download/v${version}/esp32.bin"
zip_url="https://github.com/ESPresense/ESPresense/archive/refs/tags/v${version}.zip"
wget -q "${zip_url}"
unzip -u "./$(basename ${zip_url})" -d "./" >/dev/null
mkdir -p "./ESPresense-${version}/.pio/build/esp32"
cd "./ESPresense-${version}"
platformio run -t nobuild --target upload
wget -q "${bin_url}" -O "./.pio/build/esp32/firmware.bin"

platformio run -t nobuild --target upload --upload-port <ESP32-IP>
```

If you have to downgrade multiple modules, make sure you remove the sources files and use new ones with every board...
not sure why but when I tried to re-use platformio always tried to connect to local /tty/USBXYZ where no module was connected

**Comment by gnome4711** [comment]
Thank you!

Sounds great!
